File: Dart/Sort/Gnome_Sort.dart
/*The gnome sort is a sorting algorithm which is similar to insertion sort in that it works with one item at a time but gets the item to the proper place by a series of swaps, similar to a bubble sort. The algorithm finds the first place where two adjacent elements are in the wrong order and swaps them.
*/

/*
 Input
array = [10, 34, 6, 323, 7]

// Ouput
SortedArray = [6, 7, 10, 34, 323]
Time complexities:
	Worst Case :O(n2) 
  Best Case :O(n)   
  Average Case:O(n2)
Space Complexity: Auxilary Space:O(1)  
*/

//Function sort the array using gnome sort
void gnomeSort(List arr, var n) {
  if (arr == null || n == 0) return;
  int first = 1;
  int second = 2;

  while (first < n) {
    if (arr[first - 1] <= arr[first]) {
      first = second;
      second++;
    } else {
      int temp = arr[first - 1];
      arr[first - 1] = arr[first];
      arr[first] = temp;
      first -= 1;
      if (first == 0) {
        first = 1;
        second = 2;
      }
    }
  }
}

void main() {
  //Get the array
  List arr = [10, 34, 6, 323, 7];

  //Get size of the array
  int n = arr.length;

  //Sorting of array using gnome sort
  gnomeSort(arr, n);

  //print the result
  print(arr.toString());
}

File: Python/sort/gnome_sort.py
"""
Gnome Sort algorithm implementation in Python3

Introduction:

It is a simple exchange sorting algorithm similar to insertion sort,
except that it moves an element to its proper place by a series of swaps,
as in bubble sort. In each iteration the algorithm finds the first place,
where two adjacent elements are in the wrong order and swaps them.
Afterwards it compares the swapped element with previous elements and
swaps them if they are in the wrong order. If adjacent elements are in
proper order or the end is reached, it starts a new iteration.
"""

def gnomeSort(nums):
    for i in range(1, len(nums)):
        while (i != 0 and nums[i] < nums[i - 1]):
            nums[i - 1], nums[i] = nums[i], nums[i - 1]
            i -= 1

user_input = input("Enter the numbers separated by a comma: ").strip()

# in one line: nums = [int(item) for item in user_input.split(',')]
nums = user_input.split(',')
sequence = []
for item in nums:
    sequence.append(int(item))

gnomeSort(nums)

print("Sorted nums of numbers using gnome sort method:", nums)

"""
Sample input:
Enter the numbers separated by a comma: 78,45,67,21,90

Sample output:
Sorted nums of numbers using gnome sort method: [21, 45, 67, 78, 90]

Best case time complexity: O(n)
Worst case time complexity: O(n^2)
Space complexity: O(1)
"""
